includes/amp-canonical-actions.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+require_once( AMP__DIR__ . '/includes/utils/class-amp-dom-utils.php' );
+require_once( AMP__DIR__ . '/includes/utils/class-amp-html-utils.php' );
+require_once( AMP__DIR__ . '/includes/utils/class-amp-string-utils.php' );
+
+require_once( AMP__DIR__ . '/includes/class-amp-content.php' );
+
+require_once( AMP__DIR__ . '/includes/sanitizers/class-amp-blacklist-sanitizer.php' );
+require_once( AMP__DIR__ . '/includes/sanitizers/class-amp-img-sanitizer.php' );
+require_once( AMP__DIR__ . '/includes/sanitizers/class-amp-video-sanitizer.php' );
+require_once( AMP__DIR__ . '/includes/sanitizers/class-amp-iframe-sanitizer.php' );
+require_once( AMP__DIR__ . '/includes/sanitizers/class-amp-audio-sanitizer.php' );
+
+require_once( AMP__DIR__ . '/includes/embeds/class-amp-twitter-embed.php' );
+require_once( AMP__DIR__ . '/includes/embeds/class-amp-youtube-embed.php' );
+require_once( AMP__DIR__ . '/includes/embeds/class-amp-gallery-embed.php' );
+require_once( AMP__DIR__ . '/includes/embeds/class-amp-instagram-embed.php' );
+require_once( AMP__DIR__ . '/includes/embeds/class-amp-vine-embed.php' );
+require_once( AMP__DIR__ . '/includes/embeds/class-amp-facebook-embed.php' );
+
+function build_post_content() {
+
+	$post = get_post();
+
+	$content_max_width = 1200;
+	if ( isset( $GLOBALS['content_width'] ) && $GLOBALS['content_width'] > 0 ) {
+		$content_max_width = $GLOBALS['content_width'];
+	}
+
+	$amp_content = new AMP_Content( $post->post_content,
+		array(
+			'AMP_Twitter_Embed_Handler' => array(),
+			'AMP_YouTube_Embed_Handler' => array(),
+			'AMP_Instagram_Embed_Handler' => array(),
+			'AMP_Vine_Embed_Handler' => array(),
+			'AMP_Facebook_Embed_Handler' => array(),
+			'AMP_Gallery_Embed_Handler' => array(),
+		),
+		array(
+			 'AMP_Blacklist_Sanitizer' => array(),
+			 'AMP_Img_Sanitizer' => array(),
+			 'AMP_Video_Sanitizer' => array(),
+			 'AMP_Audio_Sanitizer' => array(),
+			 'AMP_Iframe_Sanitizer' => array(
+				 'add_placeholder' => true,
+			 ),
+		),
+		array(
+			'content_max_width' => $content_max_width,
+		)
+	);
+
+	return $amp_content;
+
+}
+
+// Generate the AMP post content early on (runs the_content filters but skips our filter below)
+$GLOBALS['amp_content'] = build_post_content();
```

Comment:
I haven't touched PHP in a while – can you point me to a better way of doing this?